### ISSUE TITLE
SetFitHead incompatible with multi-label classification

### ISSUE THREAD

**Issue by OskarLiew** [huggingface/setfit#211]
Hello!

I want to use `SetFitHead` in a multi-label setting but there seems to be some problems. It works well with the logistic regression head, but not with the differentiable head.

The following code yields an error
```python
dataset = load_dataset()  # "text" and "label" keys. Labels are "multiple-hot" encoded floats, from `sklearn.preprocessing.MultiLabelBinarizer`

st_model = SentenceTransformer(modules=[word_embedding_model, pooling_model])
head = SetFitHead(
    st_model.get_sentence_embedding_dimension(),
    len(dataset["train"][0]["label"]),
)
setfitmodel = SetFitModel(st_model, head, multi_target_strategy="one-vs-all")

trainer = SetFitTrainer(
    model=setfitmodel,
    train_dataset=dataset["train"],
    eval_dataset=dataset["test"],
    loss_class=losses.CosineSimilarityLoss,
    metric=compute_metrics,
)
trainer.freeze()
trainer.train()

trainer.unfreeze()
trainer.train(
    num_epochs=25,
    batch_size=16,
    body_learning_rate=1e-5,
    learning_rate=1e-2,
    l2_weight=0.0,
)
```

When running this code, the dataloader in `setfit.train` returns long labels rather than float(they get converted by the `Dataloader` even though `SetFitDataset.y` is the correct type), leading to an exception in `torch.nn.CrossEntropyLoss`. The issue could be fixed by converting the values to floats before calculating loss.

However, even after that fix, classification is not made in a multi-label fashion for the following reasons:
- Softmax activation function; should be element-wise sigmoid for multilabel classification
- Argmax of output in `SetFitHead.predict` if `out_features > 1`

I'd like a heads-up if I am doing something wrong, otherwise I could submit a PR to make `SetFitHead` work with multi-label classification.

**Comment by tomaarsen** [comment]
cc: @blakechi as he introduced the differentiable head. 

And for @OskarLiew, I want to point your attention to #203, which intends to modify the `SetFitHead` and its loss functions. Alongside with #187 it aims to solve the issue that you described with long and float labels.

However, the other issues that you described that limit multi-label classification using the differentiable head would not be resolved with these PRs. Although I'm sure that a new PR would be appreciated in this case, there are a large number of PRs in the hopper that would cause complications during development, so beware!

- Tom Aarsen

**Comment by blakechi** [comment]
Hi @OskarLiew,

`SetFitHead` was not build for support multi-label classification initially. So it's great that you raise this up!

As @tomaarsen mentioned, there are many pending PRs right now for `SetFitHead`, which make things become complicated. But feel free to add this feature first since I have put #203 on hold. I will merge your update later in #203.

Will leave some suggestion in your PR #212!

**Comment by tomaarsen** [comment]
Implemented via #272 🎉
